tasty-plutus/src/Test/Tasty/Plutus/Internal/Monad.hs

```diff
@@ -0,0 +1,337 @@
+  [Text] ->
+  ScriptResult ->
+  m r
+deliverResult logs res = do
+  expected <- getShouldSucceed
+  case (expected, res) of
+    (_, NoOutcome) -> makeFailure =<< noOutcome
+    (False, ScriptPassed) -> makeFailure =<< unexpectedSuccess
+    (False, ScriptFailed) -> tryPass logs
+    (True, ScriptPassed) -> tryPass logs
+    (True, ScriptFailed) -> makeFailure =<< unexpectedFailure
+    (_, InternalError t) -> makeFailure <=< internalError $ t
+    (_, ParseFailed t) -> makeFailure <=< noParse $ t
+  where
+    noOutcome :: m String
+    noOutcome = do
+      state <- stateM
+      pure $
+        renderStyle ourStyle $
+          "No outcome from run"
+            $+$ state
+            $+$ ""
+            $+$ "Did you forget to use toTestValidator or toTestMintingPolicy?"
+    unexpectedSuccess :: m String
+    unexpectedSuccess = do
+      state <- stateM
+      pure $
+        renderStyle ourStyle $
+          "Unexpected success" $+$ state
+    unexpectedFailure :: m String
+    unexpectedFailure = do
+      state <- stateM
+      pure $
+        renderStyle ourStyle $
+          "Unexpected failure" $+$ state
+    internalError :: Text -> m String
+    internalError msg = do
+      state <- stateM
+      pure $
+        renderStyle ourStyle $
+          ("Internal error" <+> (text . show $ msg)) $+$ state
+    noParse :: Text -> m String
+    noParse what = do
+      state <- stateM
+      pure $
+        renderStyle ourStyle $
+          ((text . show $ what) <+> "did not parse") $+$ state
+    stateM :: m Doc
+    stateM = dumpState logs
+
+produceResult ::
+  forall (m :: Type -> Type) (p :: Purpose) (r :: Type).
+  MonadScriptTesting m p r =>
+  m r
+produceResult =
+  getScriptResult >>= \case
+    Left err -> handleError err
+    Right (logs, res) -> deliverResult logs res
+
+tryPass ::
+  forall (m :: Type -> Type) (p :: Purpose) (r :: Type).
+  MonadScriptTesting m p r =>
+  [Text] ->
+  m r
+tryPass logs =
+  getMPred >>= \case
+    Nothing -> makeSuccess =<< doPass
+    Just f ->
+      let logs' = Vector.fromList logs
+       in if f logs'
+            then makeSuccess =<< doPass
+            else makeFailure =<< didn'tLog
+  where
+    doPass :: m String
+    doPass =
+      getShouldChat <&> \case
+        Always ->
+          renderStyle ourStyle $
+            ""
+              $+$ hang "Logs" 4 (dumpLogs logs)
+        OnlyOnFail -> ""
+    didn'tLog :: m String
+    didn'tLog = do
+      state <- dumpState logs
+      pure . renderStyle ourStyle $
+        "Trace did not contain expected contents"
+          $+$ state
+
+dumpState ::
+  forall (m :: Type -> Type) (p :: Purpose) (r :: Type).
+  MonadScriptTesting m p r =>
+  [Text] ->
+  m Doc
+dumpState logs = do
+  conf <- getConf
+  sc <- getScriptContext
+  inputs <- dumpInputs
+  pure $
+    ""
+      $+$ hang "Context" 4 (valToDoc . scriptContextToValue $ sc)
+      $+$ hang "Configuration" 4 (ppDoc conf)
+      $+$ hang "Inputs" 4 inputs
```

Comment:
I am skeptical about this: it was in `Unit`'s code but not `Property`'s, since property tests use `forAllShrinkShow` to show these things instead. So it may end up printing the inputs twice.

Reply:
You are correct - there is some difference in how outputs are handled for the inputs specifically, because we perform shrinking for the property-based interface.